### ISSUE TITLE
CP Systems

### ISSUE THREAD

**Issue by rogeralsing** [ActorModelDotNet/actormodeldotnet.github.com#5]

**Comment by gabikliot** [comment]
I would really like to understand how Akka.Net (and Akka) handle/provide CP. Especially, what happens in the case of network partitions. And also **how** they provide CP (if they do) - internal mechanisms that are being used.

**Comment by rogeralsing** [comment]
Akka does not provide any such features, Akka provides Actors.
As an actor is a unique entity in time and space, it makes little sense to talk about the CAP theorem as that requires at least two nodes.
e.g. *"if you pass a message to actor A on node 1 and then read from actor A on node 2"* is just not possible as actor A will not be present on multiple places.

You would need another abstraction on top of actors in order to make CAP relevant.
e.g. a Grain, as that has other/extra properties that is not present in the original actor model.

Does this make sense?
Akka at its core is only a lib that provides the basic building blocks that you build things on top of.
e.g. Actors, Routers, Cluster etc.
What you decide to build using those blocks is up to you.

e.g. someone implemented the Raft protocol ontop of JVM Akka.
https://github.com/ktoso/akka-raft

My interpretation here is that Akka is somewhat like Lego technic bricks
![technic](https://cloud.githubusercontent.com/assets/647031/7120578/538d6f06-e20e-11e4-8b87-9bf3f2637e82.jpg)

It is possible to use Akka to build CP or CA systems, but you will have to do the grunt work yourself, but actors in its purest form gives you some guarantees that you can reason about.

While Orleans are more like a predefined larger blocks
![duplo](https://cloud.githubusercontent.com/assets/647031/7120633/b2bc8124-e20e-11e4-9c9d-3dd80d38b153.jpg)

You don't have to do as much yourself, the default scenario is already build for you (AP) so to say.
Which is all good, AFAIK one of your project goals was to bring distributed systems programming to developers that might be inexperienced with that type of systems. (?)

But you will have a very hard time modling these predefined blocks into something that they are not.
e.g. going from AP to CP.

Does my reasoning make any sense here?

**Comment by gabikliot** [comment]
So are you basically saying that Akka (and Akka.Net) is not a **distributed** platform. It is a local, one-machine building block that **can** be used in a bigger distributed platform, but by itself it is not. That sounds fare.

So the distributed platform that can be layered on top of Akka.Net, I guess there is at least one that comes somehow "prepackaged" with Akka.Net? If there is none, it's hard to see how people are using Akka.Net to build distributed systems. And people are using it right? So there must be something. So can we then ask my question regarding CP again w.r.t to this distribution layer that comes with Akka.Net?

And just to your point: we don't think it will be hard to support CP in Orleans, in addition to AP. I will take some time and write a page describing how this can be done, both at the app layer with storage help and internally within Orleans.

**Comment by rogeralsing** [comment]
No it is not a single machine block, we have Akka.Remoting and Akka.Cluster, but actors are still actors single incarnations with a specific address.

**Comment by gabikliot** [comment]
Great! 
So what is Akka.Remoting's and Akka.Cluster's take on CP?  Especially, what happens in the case of network partitions. And also how do they provide CP (if they do) - what are the internal mechanisms that are being used to provide that?

**Comment by rogeralsing** [comment]
![cp](https://cloud.githubusercontent.com/assets/647031/7131228/9ed15ca2-e27c-11e4-975f-2e8f000b4436.png)

In case of a network partition, the systems will not be able to communicate.
If you try to read a value using the `Ask` operator, the returned task will timeout.

**Comment by gabikliot** [comment]
In your example above, can you please provide more details. I am simply not familiar enough with Akka. 

1) What is the difference between ActorRef and Remote ActroRef? 
2) The write to ActorRef can come from another machine, right? Potentially from another actor on another machine, right?
3) What if the caller to Write gets a timeout exception - his call never goes by or he does not get the response back. What will it do?
4) Consider 2 cases: (a) the call Write times out on the caller (who is another actor) due to network temporal glitch and (b) the server of System A crashed. Is the caller of Write doing something different in those 2 cases? I am guessing it has to do with supervision trees. I wonder what exactly will happen, especially if supervisor is on the server where Write comes from. 

Back to your question:
> consistent view of things using the CP features of the data store?
> Isn't that a bit of a cheat in that case

The answer is yes, to both your questions. We can (don't have to, but can) rely on data store to provide CP. Call it cheating, I don't care. I call it relying on something else to provide stronger guarantees, I call it an extensible and modular architecture.

>If we plug in a data store with AP properties, and just spawn dummy actors on each system that fetch data from this store and say we are AP?

The answer to this is definitely **not**. Randomly inconsistent with no guarantees is far different from eventual guarantees. **No guarantees** (what you wrote above) is 0 guarantees; **eventual guarantees** is 100% guarantees just a bit later; **strongly consistent** (what we called CP) is 100% guarantees immediately. Fundamentally different.

**Comment by ReubenBond** [comment]
I'm confused about Akka clustering here, too. The [docs describe Akka's Cluster Singleton](http://doc.akka.io/docs/akka/snapshot/contrib/cluster-singleton.html#cluster-singleton) as AP when auto-downing is enabled, since we can see multiple instances of a singleton in the face of network partitioning, and CP (with no self-healing?) when auto-downing is disabled. The problem is pushed onto the user, correct?

How do you address an actor by name (eg the user with a given id) in Akka?

**Comment by rogeralsing** [comment]
>CP. Call it cheating, I don't care. I call it relying on something else to provide stronger guarantees, I call it an extensible and modular architecture.

But it is not a feature of the framework, this is something you can build *using* the framework, right?

**Comment by rogeralsing** [comment]
>1) What is the difference between ActorRef and Remote ActroRef? 

A remote ActorRef is pretty much just a proxy to a real ActorRef located somewhere else.
They act and look like normal actorrefs to the end user.

>2) The write to ActorRef can come from another machine, right? Potentially from another actor on another machine, right?

Yes it could, an actor or "normal" code can interact with the ActorRef.

>3) What if the caller to Write gets a timeout exception - his call never goes by or he does not get the response back. What will it do?

First; Actors are fire and forget by default, so by using `Tell` you would not get any kind of response.
Using `Ask` it's up to the target actor to send a response back.

But lets say we use `Ask` to issue a state transition on the target actor and for some reason a response does not arrive in time so the Ask Task times out.

I'm not sure what you mean by "What will it do?" You get a Task that is cancelled.
What you do with that task is up to you.
Elaborate?

>4) Consider 2 cases: (a) the call Write times out on the caller (who is another actor) due to network temporal glitch and (b) the server of System A crashed. Is the caller of Write doing something different in those 2 cases? I am guessing it has to do with supervision trees. I wonder what exactly will happen, especially if supervisor is on the server where Write comes from.

(a) you get a cancelled task, nothing breaks it is up to you to handle this.
In JVM Akka there is a `ReliableProxy` that provides at-least-once guarantees.
Not yet ported to .NET. but fairly easy to roll your own.

(b) If system A crashes, you will get the same timeout for your `Ask` operation.
And eventually your system will deem System A as unavailable.
![remoting](http://doc.akka.io/docs/akka/snapshot/_images/association_lifecycle1.png)

Supervision is involved when the target actor reports an error, if actor A on system A was remote deployed from your system onto system A. then it would pass it's parent a message when an exception occurs.

But in the case of system A crashes or becomes unavailable, the individual actors supervision trees are not involved. 
See the section "Watching remote actors": http://doc.akka.io/docs/akka/snapshot/scala/remoting.html

**Comment by sergeybykov** [comment]
I have a much more basic question. You have nodes A and B in a cluster. They both independently receive a request each that concerns user X. To process such a request they need to send a message to an actor X that represents the user. What and how guarantees that both A and B that are unaware of each other's requests will end up sending messages to a single actor X? What's the coordination mechanism to prevent A and B from creating Xa and Xb instead of a single X.

**Comment by Aaronontheweb** [comment]
@sergeybykov that issue can be solved in any number of ways:

1. Through the use of a clustered router - the consistent hash router in particular. Special types of built-in actors whose only job is to transparently route messages to other user-defined actors. Routers on either system know the full hash range belonging to each node in the network and route messages to the same place depending on the hash key value. Works conceptually the same way Cassandra does. Have the message get sent to a router first on either of the two nodes, or perhaps on the third node that made the request (you can have multiple independent services / applications running in a cluster - this is what the "roles" feature is for, and that can also be used in routing.)

2. For the case where a new node joins / leaves and the consistent hash ranges aren't caught up to the current state of the cluster yet, this is a problem that ClusterSharding can solve. Those messages will be event-sourced to one of the nodes and replicated elsewhere via a common persistence key (i.e. pulling from the same row in Cassandra or Azure table storage.) I'm not as familiar with all of the details of ClusterSharding yet - we don't have all of the dependencies it needs (Akka.Persistence, Akka.Streams) to be able to port i properly yet.

3. For cases where much higher consistency is required, you can subscribe your sender to Cluster gossip and make more explicit choices about where messages are sent. We've modeled distributed transactions this way for cases that demand high consistency.


TL;DR; - this is a problem that *should* be solved by how the message is sent and routed.

**Comment by Aaronontheweb** [comment]
> How do you address an actor by name (eg the user with a given id) in Akka?

You can use an `ActorSelection` to address an actor by its `ActorPath` - the `ActorPath` is also present on the `IActorRef` interface, so you can get the full address of any actor who's sent you a message.

**Comment by Aaronontheweb** [comment]
Akka.Cluster and Akka.Remote only make some guarantees with respect to `CP` systems out of the box - @rogeralsing described them pretty well: any remotely deployed actors between two systems that can no longer reach each other will be terminated on both ends.

that's pretty much it - otherwise Akka.Cluster is very high AP out of the box by default. 

Most of the other modules that are part of Akka's HA toolkit are designed to add tunable levels of consistency throughout the cluster. Akka.ClusterSharding, Akka.Persistence, etc... Point is to let the end-user make tradeoffs about consistency explicitly through the use of capabilities those libraries expose (like `GuaranteedDeliveryActor` etc...)

**Comment by sergeybykov** [comment]
@Aaronontheweb Is my assumption correct that the consistent hash routers by themselves cannot guarantee that A and B are talking to a single X because at any point in time the routers may have inconsistent views of the state of the cluster? Since a node (or a group of nodes) can leave/join the cluster at any time, don't you have to run a consistency protocol between routers to provide a true CP guarantee of a single X?

I don't understand how subscribing to cluster gossip can help much with consistency. Can you provide more details? If you are saying you implemented distributed transactions this way, I'm sure @philbe will want to hear about that. :-)

**Comment by Aaronontheweb** [comment]
> @Aaronontheweb Is my assumption correct that the consistent hash routers by themselves cannot guarantee that A and B are talking to a single X because at any point in time the routers may have inconsistent views of the state of the cluster

Yep. That is correct. They can't and don't try to guarantee full consistency at any given time - any time the state of the network changes those routers will experience a period of latency where the hash ranges aren't current. Takes 

> I don't understand how subscribing to cluster gossip can help much with consistency.

You use that information explicitly to communicate nodes at known addresses at the time of TransactionStart, rather than rely on a built-in Akka.NET router to do that for you since the routers abstract over that information. 

The goals of routers are throughput, configurability, and ease of use - providing a level of consistency (at the message-routing level) beyond what the `ConsistentHashRouter` can provide requires modules outside of Akka.Cluster (like ClusterSharding) or user-defined code. 

Subscribing to Cluster gossip to get that information directly is how you accomplish the latter. It's how you get closer to the metal for anything that can't be done through a built-in module.

> Since a node (or a group of nodes) can leave/join the cluster at any time, don't you have to run a consistency protocol between routers to provide a true CP guarantee of a single X?

Akka.NET will not provide any modules as part of the core project for doing distributed transactions - it's just not our area. Users of Akka have developed Paxos / RAFT / etc modules on top of it and we expect the same of Akka.NET.

> If you are saying you implemented distributed transactions this way, I'm sure @philbe will want to hear about that. :-)

For @philbe's benefit ;) 

Most of what I've been involved in doing is helping Petabridge customers figure out how to implement phase 1 of 2-phase commits - commit-request. 

At this point most of them realize that they don't really want to implement DTs (I mean, if they were trivial to implement....) and settle for something eventually consistent instead.

We gave a lightning talk at DotNetFringe showing how to model this with actors (naively, it's a 5 minute talk.) I'll post the videos and slides up when they're available.

This is just my personal bias and has zero to do with Akka.NET but I'm all for eventual consistency by default for most applications.

**Comment by sergeybykov** [comment]
@Aaronontheweb Thanks! This makes much more sense to me now. We've had a similar experience - most customers that say that they need strong consistency later admit that it's more of an illusion, and eventual consistency is a better choice for most of their scenarios.

I'm trying to reconcile what you explained with @rogeralsing' statement above:
> Akka provides Actors.
 As an actor is a unique entity in time and space

Is it fair to say that at the Akka level "a unique entity in time and space" is true for a single machine case or a stable cluster with no failures/joins/leaves, and that in a general distributed case one needs a higher/lower (depending from the point of view) level mechanism to guarantee true uniqueness of an actor with a particular name/identity?

Having skimmed through the Cluster Sharding docs, my initial impression is that it is probably subject to the same limitations as it relies on ShardCoordinator being a cluster-wide singleton with no consensus protocol mentioned other than "the oldest member among all cluster nodes or a group of nodes tagged with a specific role", where I presume different nodes may have different view of which node is the oldest at a particular point in time. Am I misinterpreting the docs?

I also glanced at Akka-Raft readme. Looks like it indeed tries to implement a true consensus protocol with some low level primitives exposed. It says it's in "early preview", but there doesn't seem to be much activity recently. Is it alive?

**Comment by rkuhn** [comment]
*This is a very interesting discussion, I would have liked to join earlier but unfortunately time did not permit it.*

@ReubenBond You are correct in your summary that the DOWNing strategy that is chosen by the user determines whether Akka Cluster is CP or AP, where auto-downing is not the only option (the classical CP options are quorum and referee, but such implementations do not currently ship with Akka). Consistent with Akka’s philosophy the user gets to choose based on the requirements of the use-case at hand, and often this boils down to a game of probabilities—neither the requirements nor the properties of the implementation are all black or white.

@gabikliot @ReubenBond You asked how an Actor is addressed in Akka. The basic abstraction is the ActorRef, and an ActorRef always references a single physical Actor location (directly or indirectly), meaning that the “name” of an Actor includes which host it is living on. As @rogeralsing explained there is no ambiguity and “CAP” terminology is irrelevant at this level of abstraction due to the lack of replication.

@sergeybykow Given a strong CP configuration of the Cluster DOWNing strategy the ClusterSingleton is actually consistent—we can prove that with our Cluster gossip (essentially a CRDT plus a version vector) there cannot be differing notions of convergence during overlapping (global) time periods within the same cluster—meaning that at any given time the set of nodes recognizing convergence is either empty or consists of nodes that agree with each other. The placement of the ClusterSingleton is based not just on the current local view of the cluster topology, it is based on the latest convergent state that was seen by all constituents. This implies that the singleton will be unavailable for a time period after its host machine crashes or becomes separated from the cluster, namely as long as it takes to conclude the node removal process and reach convergence again—with a new “oldest” member which will then start the next incarnation of the singleton (re-hydrating its state from the persistence journal, where users can again choose their favorite implementation based on required semantics).
